src/shred/shred.rs

```diff
@@ -0,0 +1,450 @@
+    }
+    
+    if remove {
+        do_remove(path, path_str, verbose).expect("Failed to remove file");
+    }
+}
+
+fn do_pass(path: &Path, generator_type: PassType,
+           given_file_size: Option<u64>, exact: bool) -> Result<(), io::Error> {
+
+    // Use the given size or the whole file if not specified
+    let size: u64 = given_file_size.unwrap_or(try!(get_file_size(path)));
+
+    let generator = BytesGenerator::new(size, generator_type, exact);
+
+    let mut file: File = try!(File::create(path));
+
+    for block in generator {
+        try!(file.write_all(&*block));
+    }
+
+    try!(file.sync_all());
+
+    Ok(())
+}
+
+fn get_file_size(path: &Path) -> Result<u64, io::Error> {
+    let file: File = try!(File::open(path));
+    let size: u64 = try!(file.metadata()).len();
+
+    Ok(size)
+}
+
+// Repeatedly renames the file with strings of decreasing length (most likely all 0s)
+// Return the path of the file after its last renaming or None if error
+fn wipe_name(orig_path: &Path, verbose: bool) -> Option<PathBuf> {
+    let file_name_len: usize = orig_path.file_name().unwrap().to_str().unwrap().len();
+    
+    let mut last_path: PathBuf = PathBuf::from(orig_path); 
+    
+    for length in (1..file_name_len + 1).rev() {
+        for name in FilenameGenerator::new(length) {
+            let new_path: PathBuf = orig_path.with_file_name(name);
+            // We don't want the filename to already exist (don't overwrite)
+            // If it does, find another name that doesn't
+            if new_path.exists() { 
+                continue; 
+            }
+            match fs::rename(&last_path, &new_path) {
+                Ok(()) => {
+                    if verbose {
+                        println!("{}: {}: renamed to {}", NAME,
+                                                          last_path.display(),
+                                                          new_path.display());
+                    }
+                   
+                    // Sync every file rename 
+                    {
+                        let new_file: File = File::create(new_path.clone()).expect("Failed to open renamed file for syncing");
+                        new_file.sync_all().expect("Failed to sync renamed file");
+                    }
+
+                    last_path = new_path;
+                    break;
+                }
+                Err(e) => {
+                    println!("{}: {}: Couldn't rename to {}: {}", NAME,
+                                                                  last_path.display(),
+                                                                  new_path.display(),
+                                                                  e);
+                    return None;
+                }
+            }
+        } // If every possible filename already exists, just reduce the length and try again
```

Comment:
There is a race condition between checking new_path.exists() and fs::rename(). new_path might have been created between the exists() call and the rename() call. Just check the error returned by rename() if it is necessary to distinguish errors.